mypy/meet.py

```diff
@@ -130,16 +131,23 @@ def narrow_declared_type(declared: Type, narrowed: Type) -> Type:
     if isinstance(declared, UnionType):
         return make_simplified_union(
             [
-                narrow_declared_type(x, narrowed)
-                for x in declared.relevant_items()
+                narrow_declared_type(d, n)
+                for d, n in product(
```

Comment:
This looks like the correct way of doing that, despite higher comp complexity. What's more mypyc-friendly - `itertools.product` or a plain listcomp with two `for` clauses?

Reply:
I feel like plain listcomp would be faster (we're consuming all the items anyways), but I would be surprised if this is a hot path because there's an `is_subtype` check -- surely that's somewhat slow!